lib/enumerable.rb

```diff
@@ -0,0 +1,115 @@
+module Enumerable
+  def my_each
+    return enum_for unless block_given?
+
+    idx = 0
+
+    arr ||= to_a
+    while idx < arr.length
+      yield(arr[idx])
+      idx += 1
+    end
+
+    arr
+  end
+
+  def my_each_with_index
+    return enum_for unless block_given?
+
+    idx = 0
+
+    while idx < length
+      yield(self[idx], idx)
+      idx += 1
+    end
+
+    self
+  end
```

Comment:
- [ ] When a block is given when self is a range, it should return the range itself after calling the given block once for each element in self. For instance, if self is a range of (5..50), this method returns the following error:
`NameError: undefined local variable or method 'length' for 5..50:Range`